runtime/reconcilers/model.go

```diff
@@ -1256,41 +1301,34 @@ func (r *ModelReconciler) executeSingle(ctx context.Context, executor *wrappedMo
 			TempDir:              tempDir,
 		})
 		if err != nil {
-			return nil, err
-		}
-		stageDuration = stageResult.ExecDuration
-
-		// We change the inputProps to be the result properties of the stage step
-		inputProps = stageResult.Properties
-
-		// Drop the stage result after the final step has executed.
-		// We do this using the same ctx, which means we may leak data in the staging connector in case of context cancellations.
-		// This is acceptable since the staging connector is assumed to be configured for temporary data.
-		defer func() {
-			err := executor.stageResultManager.Delete(ctx, stageResult)
-			if err != nil {
-				r.C.Logger.Warn("Failed to clean up staged model output", zap.String("model", self.Meta.Name.Name), zap.Error(err), observability.ZapCtx(ctx))
+			action := retryActionFromError(err, mdl.Spec.RetryIfErrorMatches)
+
+			r.C.Logger.Info("Model execution error encountered",
+				zap.String("model", self.Meta.Name.Name),
+				zap.Int("attempt", attempt),
+				zap.String("error", err.Error()),
+				zap.Strings("retry_patterns", mdl.Spec.RetryIfErrorMatches),
+				zap.Int("retry_action", int(action)),
+				observability.ZapCtx(ctx))
+
+			if action == retrier.Retry {
+				// Calculate the actual backoff duration for this attempt
+				actualBackoff := backoff
+				if mdl.Spec.RetryExponentialBackoff {
```

Comment:
I'm still confused by the changes to the actual execution implementation. I don't think the `retrier` package is helping here, as it's still mixing retry logic into the execution logic. It also seems like some unrelated changes have happened, like not logging the stage model output clean up error anymore?

Given the actual retry loop is going to be very simple, I think just wrapping the existing logic with a model-specific retry util function should be straightforward. But maybe I'm missing something. What did you think about my suggestion here? https://github.com/rilldata/rill/pull/7810#discussion_r2307434579